protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+# Purpose
+
+The purpose of this design document is to propose a mechanism for decoupling
+individual features from their hardforks in the OP Stack. This will enhance
+flexibility in feature activation, enabling granular control for testing,
+custom chain configurations, and phased rollouts while maintaining
+coordination across the superchain. The design aims to improve modularity
+without introducing excessive complexity, ultimately supporting more robust
+development and deployment practices of OP Stack software.
+
+# Background
+
+The OP Stack relies on the `RollupConfig` to specify chain parameters,
+including hardfork activation timestamps (e.g., `canyonTime`, `deltaTime`).
+These configurations are centralized in the [`superchain-registry`][scr]
+repository, as TOML files under `superchain/configs/`. For chains that are
+part of the superchain, activations are inherited from a shared timestamp
+called the `superchain_time` to ensure synchronized upgrades.
+
+Currently, when a hardfork timestamp is reached in OP Stack client software,
+all features within that hardfork (e.g., EIPs, protocol changes) activate
+simultaneously. Bundling features in this way simplifies coordination but
+limits flexibility, such as enabling individual features for testing or
+delaying specific ones without forking the config.
+
+In the past, contributors of the OP Stack have had issues surrounding
+hardfork features. For example, long before the Isthmus hardfork, the
+Operator Fee Feature was [proposed][proposed] and approved. The feature
+was considered to be a part of the upcoming Isthmus hardfork. Rust
+implementations of OP Stack components went ahead and implemented the
+feature. For months, the feature sat idle. After it was eventually merged,
+there were issues in the execution client changes for the feature. Removing
+implementations of the feature was considered, but the effort to remove it
+was not seen as worthwhile. After a while, the feature was finally
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
+In this way, features could be implemented and left in client software
+without being connected to hardfork. The key part: *no changes would
+need to be made to client software to enable the feature*, only enabling
```

Comment:
Both runtime and compile time feature toggles can be useful in different cases, but it is definitely easier to test things when using runtime feature toggles (e.g. like the suggested option in the rollup config or a CLI arg). The `IsOperatorFee()` method delegating to `IsIsthmus()` would be an example of a compile time feature and particularly when done like that would be impossible to test in the non-default configuration.

Generally, the ease of testing makes runtime feature toggles preferable, but they do have a greater risk of misconfiguration.